File: src/InsertionSort/optimized/InsertionSort.py
from ArrayGenerator import ArrayGenerator
from SortingHelper import sort_test, is_sort


class InsertionSort:
    @staticmethod
    def sort2(arr):
        for i in range(len(arr)):
            print("i", i)
            t = arr[i]
            j = i
            for j in range(0, i + 1)[::-1]:
                print("j", j)
                if j > 0:
                    if t < arr[j - 1]:
                        arr[j] = arr[j - 1]
                    else:
                        break
                # if j == 1:
                #     j = j-1
            print("change_j", j)
            # if arr == [2, 2, 3, 4, 6, 5]:
            #     print("debug")
            print(arr)
            arr[j] = t
            print("end_arr", arr)

    @staticmethod
    def sort(data_list):
        for i in range(len(data_list)):  # [0,n)
            # print(i)
            for j in range(1, i + 1)[::-1]:
                if data_list[j] < data_list[j - 1]:
                    data_list[j - 1], data_list[j] = data_list[j], data_list[j - 1]


if __name__ == '__main__':
    test_list = [2, 4, 6, 3, 1, 5]
    InsertionSort.sort2(test_list)
    raise ValueError
    print(test_list)
    InsertionSort.sort(test_list)
    print(test_list)
    input_data = ArrayGenerator.generate_random_array(10000, 10000)
    import copy
    import time

    input_data2 = copy.deepcopy(input_data)
    start = time.time()
    InsertionSort.sort(input_data)
    end = time.time()
    is_sort(input_data)
    print(f'sort: {end - start} s')

    start = time.time()
    InsertionSort.sort2(input_data2)
    end = time.time()
    is_sort(input_data2)
    print(f'sort2: {end - start} s')

    """
    [1, 2, 3, 4, 5, 6]
    [1, 2, 3, 4, 5, 6]
    sort: 9.23288869857788 s
    sort2: 4.863631963729858 s
    """

    data_size = [10000, 100000]

    for size in data_size:
        input_data = ArrayGenerator.generate_random_array(size, size)
        sort_test("InsertionSort", input_data)
    """
    java:
    SelectionSort , n = 10000 : 0.103108 s
    SelectionSort , n = 100000 : 10.020195 s
    InsertionSort , n = 10000 : 0.157813 s
    InsertionSort , n = 100000 : 15.633680 s
    
    python:
    SelectionSort, size: 10000 : 3.543562889099121 s
    SelectionSort, size: 10000 : 383.543562889099121 s
    InsertionSort, size: 10000 : 9.609786987304688 s
    
    """

File: src/InsertionSort/optimized/InsertionSort.java
package InsertionSort.optimized;

import java.util.Arrays;

public class InsertionSort {

    private InsertionSort() {
    }

    public static <E extends Comparable<E>> void sort(E[] arr) {

        for (int i = 0; i < arr.length; i++) {

            // 将 arr[i] 插入到合适的位置
//            for(int j = i; j - 1 >= 0; j --){
//                if(arr[j].compareTo(arr[j - 1]) < 0)
//                    swap(arr, j - 1, j);
//                else break;
//            }

            for (int j = i; j - 1 >= 0 && arr[j].compareTo(arr[j - 1]) < 0; j--)
                swap(arr, j - 1, j);
        }
    }

    public static <E extends Comparable<E>> void sort2(E[] arr) {

        for (int i = 0; i < arr.length; i++) { //[0,n)
            System.out.println("i" + i);
            // 将 arr[i] 插入到合适的位置
            E t = arr[i];
            int j;
            for (j = i; j - 1 >= 0; j--) { // [0,i]
                System.out.println("j" + j);
                if (t.compareTo(arr[j - 1]) < 0) {
                    arr[j] = arr[j - 1];
                } else {
                    break;
                }

            }
            System.out.println("change_j" + j);
            for (int x=0; x < arr.length; x++){
                System.out.print(arr[x]+",");
            }
            System.out.println("");
            arr[j] = t;
            System.out.println("end_arr");
            for (int x=0; x < arr.length; x++){
                System.out.print(arr[x]+",");
            }
            System.out.println("");
        }
    }

    private static <E> void swap(E[] arr, int i, int j) {

        E t = arr[i];
        arr[i] = arr[j];
        arr[j] = t;
    }

    private static <E extends Comparable<E>> boolean isSorted(E[] arr) {

        for (int i = 1; i < arr.length; i++)
            if (arr[i - 1].compareTo(arr[i]) > 0)
                return false;
        return true;
    }

    public static void main(String[] args) {

//        int[] dataSize = {10000, 100000};
        int[] dataSize = {10000};
        for (int n : dataSize) {
            Integer[] arr = ArrayGenerator.generateRandomArray(n, n);
            Integer[] arr2 = Arrays.copyOf(arr, arr.length);

            Integer[] arrTest = {2, 4, 6, 3, 1, 5};
            InsertionSort.sort2(arrTest);
            for (int x=0; x < arrTest.length; x++){
                System.out.print(arrTest[x]+",");
            }
//            SortingHelper.sortTest("InsertionSort", arr);
//            SortingHelper.sortTest("InsertionSort2", arrTest);
        }
        /**
         * InsertionSort , n = 10000 : 0.180320 s
         * InsertionSort2 , n = 10000 : 0.146157 s
         * InsertionSort , n = 100000 : 17.242725 s
         * InsertionSort2 , n = 100000 : 11.551383 s
         */
    }
}
